kernel/preemption/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//! Manages preemption on a per-CPU core basis.
+//! 
+//! Supports enabling and disabling preemption for the purpose of 
+//! safe task state management, e.g., through preemption-safe locks.
+
+#![no_std]
+
+extern crate alloc;
+
+use core::sync::atomic::{AtomicU8, Ordering};
+
+/// Theseus uses a `u8` to hold each CPU core's ID (apic ID),
+/// so the maximum number of cores it supports is `u8::MAX` (256).
+const MAX_CPU_CORES: usize = u8::MAX as usize;
+
+const ATOMIC_U8_ZERO: AtomicU8 = AtomicU8::new(0);
+
+/// The per-core preemption count, indexed by a CPU core's APIC ID.
+/// 
+/// If a CPU's count is `0`, preemption is enabled.
+/// If a CPU's count is greater than `0`, preemption is disabled.
+static PREEMPTION_COUNT: [AtomicU8; MAX_CPU_CORES] = [ATOMIC_U8_ZERO; MAX_CPU_CORES];
```

Comment:
```suggestion
static PREEMPTION_COUNT: [u8; MAX_CPU_CORES] = [u8; MAX_CPU_CORES];
```
`PREEMPTION_COUNT[apic_id]` is only accessed by the current thread so it's ok if it's not atomic.

Any reads or writes will have to be marked unsafe, but we won't need to check for overflows.